snuba/datasets/dataset.py

```diff
@@ -28,58 +28,101 @@ def get_min_columns(self) -> Sequence[str]:
+    - dataset query processing. A series of QueryProcessors are applied to the
+      query before deciding which Storage to use. These processors are defined
+      by the dataset
+    - the Storage to run the query onto is selected and the query is transformed
+      into a Storage Query. This is done by a StorageQueryPlanBuilder. This object
+      produces a plan that includes the Query contextualized on the storage/s, the
+      list of processors to apply and the strategy to run the query (in case of
+      any strategy more complex than a single DB query like a split).
+    - storage query processing. A second series of QueryProcessors are applied
+      to the query. These are defined by the storage.
+
+    The architecture of the Dataset is divided in two layers. The highest layer
+    provides the logic we use to deal with the data model. (writers, query processors,
+    query planners, etc.). The lowest layer incldues simple objects that defines
+    the query itself (Query, Schema, RelationalSource). The lop layer object access and
+    manipulate the lower layer objects.
     """
 
     def __init__(
         self,
-        dataset_schemas: DatasetSchemas,
         *,
-        table_writer: Optional[TableWriter] = None,
+        storages: Sequence[Storage],
+        query_plan_builder: StorageQueryPlanBuilder,
+        abstract_column_set: ColumnSet,
+        writable_storage: Optional[TableStorage],
     ) -> None:
-        self.__dataset_schemas = dataset_schemas
-        self.__table_writer = table_writer
+        self.__storages = storages
+        self.__query_plan_builder = query_plan_builder
+        self.__abstract_column_set = abstract_column_set
+        self.__writable_storage = writable_storage
 
-    def get_dataset_schemas(self) -> DatasetSchemas:
+    def get_extensions(self) -> Mapping[str, QueryExtension]:
         """
-        Returns the collections of schemas for DDL operations and for
-        query.
-        See TableWriter to get a write schema.
+        Returns the extensions for this dataset.
+        Every extension comes as an instance of QueryExtension.
+        The schema tells Snuba how to parse the query.
+        The processor actually does query processing for this extension.
         """
-        return self.__dataset_schemas
+        raise NotImplementedError("dataset does not support queries")
 
-    def can_write(self) -> bool:
+    def get_query_processors(self) -> Sequence[QueryProcessor]:
         """
-        Returns True if this dataset has write capabilities
+        Returns a series of transformation functions (in the form of QueryProcessor objects)
+        that are applied to queries after parsing and before running them on the storage.
+        These are applied in sequence in the same order as they are defined and are supposed
+        to be stateless.
+        """
+        return []
+
+    def get_abstract_columnset(self) -> ColumnSet:
```

Comment:
This represents the Schema the Snuba Query (not the storage query) can be ran against. Not that anyone cares during query processing today. I called it abstract ColumnsSet because it returns a ColumnSet and not a Schema, .... because a Schema is something that has DDL information associated.
Maybe we should reshuffle the names around the Schema api.